File: generated_code/cpp/solution_2299_2.cpp
#include <iostream>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    /**
     * Algorithm: In-place Two-Pointer approach.
     * We use a 'writer' pointer to modify the existing nodes of the list to save space
     * and a 'reader' pointer to traverse and sum the values between zeros.
     * 
     * Time Complexity: O(n) - We traverse the list exactly once.
     * Space Complexity: O(1) - We modify the list in-place (excluding the return result).
     */
    ListNode* mergeNodes(ListNode* head) {
        // Start 'writer' at the first node that will hold a sum (head->next).
        // The original head is a 0, so we ignore it.
        ListNode* writer = head->next;
        ListNode* reader = head->next;
        
        while (reader != nullptr) {
            int currentSum = 0;
            
            // Traverse until we hit the next 0 node.
            // Since the problem guarantees no two consecutive zeros and ends with a 0,
            // we will always find a sequence of non-zero nodes.
            while (reader != nullptr && reader->val != 0) {
                currentSum += reader->val;
                reader = reader->next;
            }
            
            // Assign the sum to the writer node.
            writer->val = currentSum;
            
            // Move reader to the node immediately after the 0 we just encountered.
            reader = reader->next;
            
            // If reader is null, we reached the end of the original list.
            if (reader == nullptr) {
                writer->next = nullptr;
                break;
            }
            
            // Otherwise, link the current writer to the next available slot and move forward.
            writer->next = reader;
            writer = writer->next;
        }
        
        // Return head->next because the original head was a 0.
        return head->next;
    }
};

File: generated_code/python/solution_2299_2.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Algorithm: Two Pointers (In-place modification)
        Time Complexity: O(n) - We visit each node in the linked list exactly once.
        Space Complexity: O(1) - We reuse the existing nodes to build the result list, 
                                 modifying it in-place without extra memory allocation.
        """
        # 'modify' tracks the position where the next summed value will be stored.
        # 'curr' traverses through the rest of the list to calculate the sums.
        modify = head.next
        curr = modify
        
        while curr is not None:
            current_sum = 0
            # Accumulate values until encountering the next 0 node
            while curr and curr.val != 0:
                current_sum += curr.val
                curr = curr.next
            
            # Store the accumulated sum in the 'modify' node
            modify.val = current_sum
            
            # Advance 'curr' past the 0 node to start the next segment
            if curr:
                curr = curr.next
                
            # Link the current modified node to the next segment's modify node
            modify.next = curr
            modify = modify.next
            
        return head.next